File: go/numeric.go
package main

func Itoa(v []int, increment int) {
	for i := range v {
		v[i] = increment
		increment++
	}
}

func Accumulate(v []int, init int, op func(int, int) int) int {
	for _, e := range v {
		init = op(init, e)
	}
	return init
}

func InnerProduct(a []int, b []int, init int) int {
	var size int
	if len(a) > len(b) {
		size = len(b)
	} else {
		size = len(a)
	}
	for i := 0; i < size; i++ {
		init += (a[i] * b[i])
	}
	return init
}

func AdjacentDifference(v []int) []int {
	size := len(v)
	res := append([]int{}, v...)
	for i := 1; i < size; i++ {
		res[i] = v[i] - v[i-1]
	}
	return res
}

func PartialSum(v []int) []int {
	size := len(v)
	res := append([]int{}, v...)
	for i := 1; i < size; i++ {
		res[i] = v[i] + res[i-1]
	}
	return res
}

File: ruby/numeric.rb
# Implementation of numeric algorithms

module NumericAlgorithm

def itoa(v)
  Enumerator.new do |yielder|
    value = v
    loop do
      yielder << value
      value += 1
    end
  end
end

def accumulate(seq, init, &block)
  seq.reduce(init, &block)
end

def inner_product(seq1, seq2, init)
  raise ArgumentError.new('second sequence should be as long as the first one'
                         ) if seq1.size != seq2.size
  seq1.zip(seq2).reduce(init){|x, y| x + y[0] * y[1]}
end

def adjacent_difference(seq)
  res = seq[0...seq.size-1].unshift(0)
  seq.each_with_index{|v, i| res[i] = v - res[i]}
  res
end

def partial_sum(seq)
  res = seq[0...seq.size-1].unshift(seq[0])
  seq.each_with_index{|v, i| res[i] = res[i-1] + v if i >= 1}
  res
end

end
